ANDROID: x86_64: Set default max-page-size to 16kB

Kalesh-Singh opened the pull request:
Android now supports both 4kB and 16kB page sizes. The vast majority of android apps are developed on x86_64 machines. In order to provide emulators that support larger page sizes, Android emulates the page-size in x86_64 to support testing apps for large page size support.

For this reason, update Android x86_64 ELFs default max-page-size to 16384 to support both 4kB and 16kB page-size devices.

Increase max-page-size raises concerns of increased disk space and extra VMA slab memory.

In Android, RO partitions use sparse images, so that the holes on ELFs don't allocate blocks on disk; and PackageManager ensures to punch holes in ELF-paddings on the /data partition when apps are installed.

Extra VMA slab memory is addressed by the bionic loader, which extends segment VMAs to cover the gaps between consecutive segment mappings, to avoid the extra VMAs needed for the gap PROT_NONE mappings (---p). This optimization is done in the crt_pad_segment note [1] is present in the ELF.

[1] https://cs.android.com/android/platform/superproject/main/+/189e480390ef13199d59e1fb54078e8b78ea6f79:bionic/libc/arch-common/bionic/crt_pad_segment.S

Commit:
ANDROID: x86_64: Set default max-page-size to 16kB

Android now supports both 4kB and 16kB page sizes. The vast majority
of android apps developed on x86_64 machines. In order to provide
emulators that support larger page sizes, Android emulates the page-size
in x86_64 to support testing apps for large page size support.

For this reason, update Android x86_64 ELFs default max-page-size to
16384 to support both 4kB and 16kB page-size devices.

It raises the concern of increased disk space and extra VMA slab memory.

In Android, RO partitions use sparse images, so that the holes on ELFs
don't allocate blocks on disk; and PackageManger ensures to punch
holes in ELF-paddings on the /data partition when an app is installed.

Extra VMA slab memory is addressed by the bionic loader, which extends
segment VMAs to cover the gaps between consecutive segment mappings, to
avoid the extra VMAs needed for the gap PROT_NONE mappings (---p). This
optimization is done in the crt_pad_segment note [1] is present in the ELF.

[1] https://cs.android.com/android/platform/superproject/main/+/189e480390ef13199d59e1fb54078e8b78ea6f79:bionic/libc/arch-common/bionic/crt_pad_segment.S

Signed-off-by: Kalesh Singh <kaleshsingh@google.com>

pirama-arumuga-nainar commented:
Can you add a test similar to #70251?

Kalesh-Singh committed:
Merge branch 'llvm:main' into x86_64-16k

Kalesh-Singh committed:
Merge branch 'llvm:main' into x86_64-16k

Commit:
Add max-page-size test for x86_64/i386

Kalesh-Singh commented:
> Can you add a test similar to #70251?

Added tests for x86[_64], thanks

MaskRay commented on clang/test/Driver/android-link.cpp:
```diff
@@ -17,10 +17,20 @@
-// RUN: FileCheck -check-prefix=MAX-PAGE-SIZE-AARCH64 < %t %s
+// RUN: FileCheck -check-prefix=MAX-PAGE-SIZE-16KB < %t %s
+//
+// RUN: %clang -### -target i386-none-linux-android %s 2>&1 | \
```
`--target=` for new tests. `-target ` has been deprecated since ~3.4

Kalesh-Singh replied:
Updated tests to use `--target=`

MaskRay commented on clang/test/Driver/android-link.cpp:
```diff
@@ -17,10 +17,20 @@
+//
+// RUN: %clang -### -target i386-none-linux-android %s 2>&1 | \
+// RUN:   FileCheck --check-prefix=NO-MAX-PAGE-SIZE-16KB %s
+//
```
Don't add `^//$` lines

Kalesh-Singh replied:
Removed `^//$` lines. Thanks for reviews.

Commit:
Refactor clang/test/Driver/android-link.cpp

Remove ^//$ lines and use --target= instead of -target which has been
deprecated.

Signed-off-by: Kalesh Singh <kaleshsingh@google.com>
